### PR TITLE
[GDI32_APITEST] Create StretchBlt Regression Test

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Add new Regression Unit Test for StretchBlt functions to gdi32_apitest.

_Added new code to gdi32_apitest for testing StretchBlt functions._

JIRA issue: [CORE-16642](https://jira.reactos.org/browse/CORE-16642)

### CODE REVIEW

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    DeleteObject(bmpDst);
+
+    memset(&biDst, 0, sizeof(BITMAPINFO));              // Clear our Bitmap Into to all zeroes
+
+    biDst.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);  // Set our Size to the header size
+    biDst.bmiHeader.biWidth = 2;                        // Set our Width to 2
+    biDst.bmiHeader.biHeight = 2;  // bottom-up         // Set our Height to 2 that's positive (bottom-up)
+    biDst.bmiHeader.biPlanes = 1;                       // Set out planes to 1 (1 required by Windows)
+    biDst.bmiHeader.biBitCount = 32;                    // Set out BitCount to 32 (Full Color)
+    biDst.bmiHeader.biCompression = BI_RGB;             // Set our Compression to BI_RBG (uncompressed)
```
Comment (Mas4hmad) [inline modules/rostests/apitests/gdi32/StretchBlt.c:65]: Typo in comment

Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c:65]: Corrected. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    expected[0] = 255*256*256;
+    expected[1] = 127*256*256*256;
+    expected[2] = 255;
+    expected[3] = 255*256;
+
+    ok(expected[0] == dstBuffer[0], "StretchBlt expected { %08X } got { %08X }\n",
+        expected[0], dstBuffer[0] );
+
+    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
+        expected[1], dstBuffer[1] );
```
Comment (binarymaster) [inline modules/rostests/apitests/gdi32/StretchBlt.c:103]:
```suggestion
    ok(expected[0] == dstBuffer[0], "StretchBlt expected { %08X } got { %08X }\n",
        expected[0], dstBuffer[0]);

    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
        expected[1], dstBuffer[1]);
```

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    expected[0] = 255*256;
+    expected[1] = 255;
+    expected[2] = 127*256*256*256;
+    expected[3] = 255*256*256;
+
+    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
+        expected[1], dstBuffer[1] );
+
+    ok(expected[3] == dstBuffer[3], "StretchBlt expected { %08X } got { %08X }\n",
+        expected[3], dstBuffer[3] );
```
Comment (binarymaster) [inline modules/rostests/apitests/gdi32/StretchBlt.c:90]: No space needed before `)`
```suggestion
    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
        expected[1], dstBuffer[1]);

    ok(expected[3] == dstBuffer[3], "StretchBlt expected { %08X } got { %08X }\n",
        expected[3], dstBuffer[3]);
```

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    /* Flip Top to Bottom  on Source */
+    StretchBlt(hdcDst, 0, 0, 2, 2, hdcSrc, 0, 1, 2, -2, SRCCOPY );
+    expected[0] = 255*256*256;
+    expected[1] = 127*256*256*256;
+    expected[2] = 255;
+    expected[3] = 255*256;
```
Comment (binarymaster) [inline modules/rostests/apitests/gdi32/StretchBlt.c:97]:
```suggestion
    /* Flip Top to Bottom on Source */
    StretchBlt(hdcDst, 0, 0, 2, 2, hdcSrc, 0, 1, 2, -2, SRCCOPY);
    expected[0] = 0x00FF0000;
    expected[1] = 0x7F000000;
    expected[2] = 0x000000FF;
    expected[3] = 0x0000FF00;
```

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    /* Flip Left to Right and Top To Bottom on both Source and Destination  */
+    StretchBlt(hdcDst, 1, 1, -2, -2, hdcSrc, 1, 1, -2, -2, SRCCOPY );
+    expected[0] = 127*256*256*256;
+    expected[1] = 255*256*256;
+    expected[2] = 255*256*256;
+    expected[3] = 255;
+
+    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
+        expected[1], dstBuffer[1] );
```
Comment (binarymaster) [inline modules/rostests/apitests/gdi32/StretchBlt.c:129]:
```suggestion
    /* Flip Left to Right and Top to Bottom on Destination */
    StretchBlt(hdcDst, 1, 1, -2, -2, hdcSrc, 0, 0, 2, 2, SRCCOPY);
    expected[0] = 0x7F000000;
    expected[1] = 0x00FF0000;
    expected[2] = 0x0000FF00;
    expected[3] = 0x000000FF;

    StretchBlt(hdcDst, 0, 0, 2, 2, hdcSrc, 1, 1, -2, -2, SRCCOPY);
    expected[0] = 0x7F000000;
    expected[1] = 0x00FF0000;
    expected[2] = 0x0000FF00;
    expected[3] = 0x000000FF;

    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
        expected[1], dstBuffer[1]);

    /* Flip Left to Right and Top to Bottom on both Source and Destination */
    StretchBlt(hdcDst, 1, 1, -2, -2, hdcSrc, 1, 1, -2, -2, SRCCOPY);
    expected[0] = 0x7F000000;
    expected[1] = 0x00FF0000;
    expected[2] = 0x00FF0000;
    expected[3] = 0x000000FF;

    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
        expected[1], dstBuffer[1]);
```

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    /* Flip Left to Right and Top To Bottom on both Source and Destination  */
+    StretchBlt(hdcDst, 1, 1, -2, -2, hdcSrc, 1, 1, -2, -2, SRCCOPY );
+    expected[0] = 127*256*256*256;
+    expected[1] = 255*256*256;
+    expected[2] = 255*256*256;
```
Comment (binarymaster) [inline modules/rostests/apitests/gdi32/StretchBlt.c:125]: Is it correct that these values are equal?

**Review by binarymaster — CHANGES_REQUESTED** [review]

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    DeleteObject(bmpDst);
+
+    memset(&biDst, 0, sizeof(BITMAPINFO));              // Clear our Bitmap Into to all zeroes
+
+    biDst.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);  // Set our Size to the header size
+    biDst.bmiHeader.biWidth = 2;                        // Set our Width to 2
+    biDst.bmiHeader.biHeight = 2;                       // Set our Height to 2 that's positive (bottom-up)
+    biDst.bmiHeader.biPlanes = 1;                       // Set out planes to 1 (1 required by Windows)
+    biDst.bmiHeader.biBitCount = 32;                    // Set out BitCount to 32 (Full Color)
+    biDst.bmiHeader.biCompression = BI_RGB;             // Set our Compression to BI_RBG (uncompressed)
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: I left two that are somewhat unique. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+    bmpSrc = CreateDIBSection(hdcSrc, &biSrc, DIB_RGB_COLORS, (void**)&srcBuffer, NULL, 0);
+    oldSrc = SelectObject(hdcSrc, bmpSrc);
+    bmpDst = CreateDIBSection(hdcDst, &biDst, DIB_RGB_COLORS, (void**)&dstBuffer, NULL, 0);
+    oldDst = SelectObject(hdcDst, bmpDst);
+
+    srcBuffer[0] = 255;             // BLUE// stored beginning bottom left
+    srcBuffer[1] = 255*256;         // GREEN
+    srcBuffer[2] = 255*256*256;     // RED
+    srcBuffer[3] = 127*256*256*256; // BLACK
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: I have converted the notation and shown that I was using hexadecimal printouts and the alpha of 0x7F000000 was easy to recognize. When I used 255 * 256 * 256 * 256 the GCC compiler gave me an overflow on INT error, so I just left off the highest bit, but using the hex notation the GCC compiler did not complain, so I have made it 0xFF000000 now. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+    /* Flip Left to Right on Source */
+    StretchBlt(hdcDst, 0, 0, 2, 2, hdcSrc, 1, 0, -2, 2, SRCCOPY );
+    expected[0] = 255*256;
+    expected[1] = 255;
+    expected[2] = 127*256*256*256;
+    expected[3] = 255*256*256;
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: Fixed. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 2 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: Fixed. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 3 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: Fixed. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 4 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: Fixed. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 5 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: Fixed. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 6 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: These are the actual results on Windows Server 2003 SP2  and Win7 when tested. I don't understand why either.

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    expected[0] = 0xFF000000;
+    expected[1] = 0x00FF0000;
+    expected[2] = 0x0000FF00;
+    expected[3] = 0x000000FF;
+
+    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
+        expected[1], dstBuffer[1]);
+
+    /* Flip Left to Right and Top to Bottom on both Source and Destination */
+    StretchBlt(hdcDst, 1, 1, -2, -2, hdcSrc, 1, 1, -2, -2, SRCCOPY );
```
Comment (binarymaster) [inline modules/rostests/apitests/gdi32/StretchBlt.c:116]:
```suggestion
    StretchBlt(hdcDst, 1, 1, -2, -2, hdcSrc, 1, 1, -2, -2, SRCCOPY);
```

File: modules/rostests/apitests/gdi32/StretchBlt.c [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+#define WIN32_NO_STATUS
+#include "windef.h"
+#include "winbase.h"
+#include "winerror.h"
+#include "wingdi.h"
+#include "winuser.h"
+#include "mmsystem.h"
+#include "wine/winternl.h"
+#include "wine/test.h"
+
```
Comment (binarymaster) [inline modules/rostests/apitests/gdi32/StretchBlt.c:22]: Not sure all these header files are required, especially `ntstatus.h` and `mmsystem.h`, maybe it's possible to minimize inclusions.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 10 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: Fixed. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 11 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c:20]: The "wine/test.h" is where the "OK" test comes from, so it is needed. I removed the other two. Thanks.

File: modules/rostests/apitests/gdi32/StretchBlt.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (learn-more) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: are the colors in the `RGB` macro in the wrong order?

[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/StretchBlt.c]: It depends on whether you mean the one in "drivers\base\bootvid\precomp.h" or "sdk\include\psdk\wingdi.h". One is good for BitBlt and the other for "x = y" (little endian) versions, so both are useful.

**Comment by katahiromz** [comment]
Creating DIB section can be done by a helper function...

**Comment by Doug-Lyons** [comment]
@katahiromz, I agree that we need many more tests than this, but CreateDIBSection or associated functions in ReactOS seems to me to be very broken at this time. So in using it to verify operation of the StretchBlt, I have seen many failures in ReactOS when in Windows they work fine. My suggestions is to try and get this function and it associated functions fixed and then look into adding helper functions and more tests.

**Review by learn-more — APPROVED** [review]
Sorry it took a while :)
Test seems good, and passes on 2k3

**Review by binarymaster — APPROVED** [review]

**Pull request merged**